### PR TITLE
change the pT3 rz chi2 calculation from linear to helix

### PR DESCRIPTION
**Pull request by YonsiG**
efficiency kept roughly the same but the fake rate of pT3 is reduced quite a bit. both because the previous cut is tuned based on pT>1GeV and now is 0.8GeV, and also the chi2 in rz plane is calculated from a helix function instead of the linear one.

### CODE REVIEW

**Comment by YonsiG** [comment]
efficiency comparison with master 
https://www.classe.cornell.edu/~ygu/www/rz_geo12_new_charge/efficiency_comparison/eff_comparison_plots__2bc0096DIRTY_explicit_cache_6a246cdDIRTY_explicit_cache_PU200/#3.1

**Comment by YonsiG** [comment]
Timing information
Screen Shot 2022-10-11 at 6 51 00 PM

**Comment by tresreid** [comment]
This is tuned for CMSSW 12? or 10?

File: SDL/PixelTriplet.cu [hunk 1]
```diff
@@ -2502,6 +2564,7 @@ __device__ bool inline SDL::runpT5DefaultAlgoPPBB(struct modules& modulesInGPU,
     float pz = segmentsInGPU.pz[pixelSegmentArrayIndex];
     float ptErr = segmentsInGPU.ptErr[pixelSegmentArrayIndex];
     float etaErr = segmentsInGPU.etaErr[pixelSegmentArrayIndex];
+    printf("%f\n",etaErr);
```
Comment (tresreid) [inline SDL/PixelTriplet.cu:2567]: can you comment this out or something?

**Comment by YonsiG** [comment]
> This is tuned for CMSSW 12? or 10?

It's under CMSSW 12 :)

File: SDL/PixelTriplet.cu (hunk 1 above)
Comment (YonsiG) [reply SDL/PixelTriplet.cu]: Thanks! Fixed in next commit

File: SDL/Event.cu [hunk 2]
```diff
@@ -725,8 +725,13 @@ __global__ void addPixelSegmentToEventKernel(unsigned int* hitIndices0,unsigned
       unsigned int outerMDIndex = rangesInGPU.miniDoubletModuleIndices[pixelModuleIndex] + 2*(tid) +1;
       unsigned int pixelSegmentIndex = rangesInGPU.segmentModuleIndices[pixelModuleIndex] + tid;
 
+//#ifdef CUT_VALUE_DEBUG
+//      addMDToMemory(mdsInGPU, hitsInGPU, modulesInGPU, hitIndices0[tid], hitIndices1[tid], pixelModuleIndex, 0,0,0,0,0,0,0,0,0,0,0,0,0,innerMDIndex);
+//      addMDToMemory(mdsInGPU, hitsInGPU, modulesInGPU, hitIndices2[tid], hitIndices3[tid], pixelModuleIndex, 0,0,0,0,0,0,0,0,0,0,0,0,0,outerMDIndex);
+//#else
```
Comment (VourMa) [inline SDL/Event.cu:731]: Why are these commented out? Are we to drop them or is this a leftover from a test?

File: SDL/TrackCandidate.cu [hunk 3]
```diff
@@ -328,14 +328,14 @@ __global__ void SDL::crossCleanpLS(struct SDL::modules& modulesInGPU, struct SDL
...
-__global__ void SDL::addpT3asTrackCandidatesInGPU(struct SDL::pixelTriplets& pixelTripletsInGPU, struct SDL::trackCandidates& trackCandidatesInGPU)
+__global__ void SDL::addpT3asTrackCandidatesInGPU(struct SDL::pixelTriplets& pixelTripletsInGPU, struct SDL::trackCandidates& trackCandidatesInGPU, struct SDL::segments& segmentsInGPU)
 {
     unsigned int nPixelTriplets = *pixelTripletsInGPU.nPixelTriplets;
 
     for(int pixelTripletIndex = blockIdx.x * blockDim.x + threadIdx.x; pixelTripletIndex < nPixelTriplets; pixelTripletIndex += blockDim.x*gridDim.x)
     {
-        if(pixelTripletsInGPU.isDup[pixelTripletIndex]) continue;
-
+//        if((!segmentsInGPU.isQuad[pixelTripletsInGPU.pixelSegmentIndices[pixelTripletIndex]]) || (pixelTripletsInGPU.isDup[pixelTripletIndex])) continue;//return;
```
Comment (VourMa) [inline SDL/TrackCandidate.cu:337]: This is commented out and, if I understand correctly, this was the only place that `segmentsInGPU` were used in this kernel. If this is the case, should we drop this and also the extra argument in the kernel call?

File: SDL/PixelTriplet.cu [hunk 4]
```diff
@@ -360,52 +368,53 @@ __device__ bool SDL::passPT3RPhiChiSquaredInwardsCuts(struct modules& modulesInG
...
         return chiSquared < 935179.56807f;
     }
-    else if(layer1 == 8 and layer2 == 9 and layer3 == 10)
+    else if(layer1 == 8 and layer2 == 9 and layer3 == 10) // endcap layer 2,3,4
     {
         return chiSquared < 29064.12959f;
     }
-    else if(layer1 == 8 and layer2 == 9 and layer3 == 15)
+    else if(layer1 == 8 and layer2 == 9 and layer3 == 15) // endcap layer 2,3,4, layer3->2s      
+    // why don't we have endcap 3,4,5 for 3 ps modules? that's 9,10,11
```
Comment (VourMa) [inline SDL/PixelTriplet.cu:385]: Did we answer that in a meeting? Should we discuss it and remove this comment?

File: SDL/PixelTriplet.cu [hunk 5]
```diff
@@ -813,6 +873,7 @@ __global__ void SDL::createPixelTripletsInGPUFromMapv2(struct SDL::modules& modu
 
         if(segmentsInGPU.isDup[i_pLS]) continue;
         if(segmentsInGPU.partOfPT5[i_pLS]) continue;//don't make pT3s for those pixels that are part of pT5
+//        if(!segmentsInGPU.isQuad[i_pLS]) continue;
```
Comment (VourMa) [inline SDL/PixelTriplet.cu:876]: Remove?

File: SDL/PixelTriplet.cu [hunk 6]
```diff
@@ -840,6 +901,7 @@ __global__ void SDL::createPixelTripletsInGPUFromMapv2(struct SDL::modules& modu
 
             float pixelRadius, pixelRadiusError, tripletRadius, rPhiChiSquared, rzChiSquared, rPhiChiSquaredInwards, centerX, centerY;
             bool success = runPixelTripletDefaultAlgo(modulesInGPU, rangesInGPU, mdsInGPU, segmentsInGPU, tripletsInGPU, pixelSegmentIndex, outerTripletIndex, pixelRadius, pixelRadiusError, tripletRadius, centerX, centerY, rzChiSquared, rPhiChiSquared, rPhiChiSquaredInwards);
+//                printf("%f ",rzChiSquared);
```
Comment (VourMa) [inline SDL/PixelTriplet.cu:904]: Remove?

File: code/core/write_sdl_ntuple.cc [hunk 7]
```diff
@@ -301,7 +301,7 @@ void createLowerLevelOutputBranches()
     createPixelTripletCutValueBranches();
     createQuintupletCutValueBranches();
     createPixelQuintupletCutValueBranches();
-    createTrackExtensionCutValueBranches();
+//    createTrackExtensionCutValueBranches();
```
Comment (VourMa) [inline code/core/write_sdl_ntuple.cc:304]: I guess this and some following commented out code will be dropped in another PR? Or do we want to drop it here?

File: SDL/PixelTriplet.cu [hunk 8]
```diff
@@ -433,78 +442,128 @@ __device__ bool SDL::passPT3RZChiSquaredCuts(struct modules& modulesInGPU, uint1
...
+            float B=2*paraA*paraB;
+            float C=paraA*paraA-paraC*paraC;
+            //A*sin(\rho s)*sin(\rho s)+B*sin(\rho s)+C=0;
+            float sol1 = (-B+sqrt(B*B-4*A*C))/(2*A);
+            float sol2 = (-B-sqrt(B*B-4*A*C))/(2*A);
+            float solz1 = asin(sol1)/rou*Pz/p+z1;
+            float solz2 = asin(sol2)/rou*Pz/p+z1;
+            float diffz1 = fabs(solz1-zsi)*100;
+            float diffz2 = fabs(solz2-zsi)*100;
+            diffz = min(diffz1,diffz2);
```
Comment (VourMa) [inline SDL/PixelTriplet.cu:559]: Thanks a lot for these calculations, Yanxi. I think it might be a good idea to have a comment in the code pointing to slides 19-21 in your presentation (https://indico.cern.ch/event/1185895/contributions/4982756/attachments/2526561/4345805/helix%20pT3%20summarize.pdf), so that we know how and why these all these computations are there.

File: SDL/PixelTriplet.cu [hunk 9]
```diff
@@ -433,78 +442,128 @@ __device__ bool SDL::passPT3RZChiSquaredCuts(struct modules& modulesInGPU, uint1
...
+        }
+
+        float diffr_f = diffr;
+        float diffz_f = diffz;
+
+        residual = moduleSubdet == SDL::Barrel ? diffz_f : diffr_f ;
+
```
Comment (VourMa) [inline SDL/PixelTriplet.cu:566]: Why have all of these and not directly set
```
residual = fabs(rtsi-sqrt(x*x+y*y))*100;
```
in the `moduleSubdet == SDL::Endcap` case and
```
residual = min(diffz1,diffz2);
```
in the `moduleSubdet == SDL::Barrel` case?

File: SDL/PixelTriplet.cu [hunk 10]
```diff
@@ -433,78 +442,128 @@ __device__ bool SDL::passPT3RZChiSquaredCuts(struct modules& modulesInGPU, uint1
...
     //hardcoded array indices!!!
     float RMSE = 0;
+
+    float Pt=pixelSegmentPt, Px=pixelSegmentPx, Py=pixelSegmentPy, Pz=pixelSegmentPz;
+    int charge=pixelSegmentCharge;
+    float CenterX=pixelG/100, CenterY=pixelF/100, CircleRadius=pixelRadiusPCA/100; // for calculating initial state corrections
+    float x0 = xPix[0]/100,x1 = xPix[1]/100;
+    float y0 = yPix[0]/100,y1 = yPix[1]/100;
+    float z0 = zPix[0]/100,z1 = zPix[1]/100;
+    float r0 = rtPix[0]/100,r1 = rtPix[1]/100;
```
Comment (VourMa) [inline SDL/PixelTriplet.cu:516]: I couldn't find where we use `CenterX`, `CenterY`, `CircleRadius`, `x0`, `y0`, `z0` and `r0`. Are they used somewhere and I missed it?

File: SDL/Event.cu (hunk 2 above)
Comment (YonsiG) [reply SDL/Event.cu:731]: yes, this not used anywhere :)

File: SDL/PixelTriplet.cu (hunk 4 above)
Comment (YonsiG) [reply SDL/PixelTriplet.cu:385]: yes, that's a leftover question when I first saw the code and has been answered

File: SDL/TrackCandidate.cu (hunk 3 above)
Comment (YonsiG) [reply SDL/TrackCandidate.cu:337]: Yes, it's the only place that use the segmentsInGPU in this function. But I propose to leave it here because this commented out line can be used if we want to control the pixel is from quad, we can remove the comment and test here? probably that could be useful later?

Comment (YonsiG) [reply SDL/TrackCandidate.cu:337]: removed

File: SDL/PixelTriplet.cu (hunk 5 above)
Comment (YonsiG) [reply SDL/PixelTriplet.cu:876]: removed

File: SDL/PixelTriplet.cu (hunk 6 above)
Comment (YonsiG) [reply SDL/PixelTriplet.cu:904]: remove

File: SDL/PixelTriplet.cu (hunk 8 above)
Comment (YonsiG) [reply SDL/PixelTriplet.cu:559]: Thanks for this suggestion! I have added the link~

File: SDL/PixelTriplet.cu (hunk 9 above)
Comment (YonsiG) [reply SDL/PixelTriplet.cu:566]: I cleaned the redundant variable

File: SDL/PixelTriplet.cu (hunk 10 above)
Comment (YonsiG) [reply SDL/PixelTriplet.cu:516]: they are no longer used, let me remove those

**Comment by YonsiG** [comment]
https://www.classe.cornell.edu/~ygu/www/rz_pixelmap_new/eff_comparison_plots__PU200_79bb1a8_explicit_cache_PU200_751bb75_explicit_cache/
This is the new efficiency comparison, master is after pulling down changing pixel map

File: code/core/write_sdl_ntuple.cc (hunk 7 above)
Comment (VourMa) [reply code/core/write_sdl_ntuple.cc:304]: I guess this will be part of Philip's upcoming clean-up PR, so let's leave it as is for now, since it doesn't break anything.

**Comment by VourMa** [comment]
Thanks a lot for the changes, @YonsiG, the PR looks great to me, both code- and performance-wise. I let @tresreid also comment on the validation and merge the PR, if everything is in order.

**Comment by tresreid** [comment]
This looks great

**Comment by YonsiG** [comment]
Thank you very much Manos and Tres!

**Comment by YonsiG** [comment]
This is the slides related to the summary work of this PR.
[helix pT3 summarize.pdf](https://github.com/SegmentLinking/TrackLooper/files/14102277/helix.pT3.summarize.pdf)
